### PR TITLE
Improve calendar connectivity

### PR DESCRIPTION
**Pull request by machria**
We modify calendar architecture to generalize connection on calendars in NextCloud and Fruux platforms. (classes are in io.github.oliviercailloux.jconfs.calendar either in src/main/java or in src/test/java)
We also tried an architecture with an interface and two different classes.
Nevertheless, we think we can merge the two different behaviour.
A try catch may need to be implemented for url property either in conference reader or in calendar online.
Multiple Unit tests have been implemented.

### CODE REVIEW

File: src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java [hunk 1]
```diff
@@ -124,7 +123,7 @@ public void display() {
 	 */
 	public void getConferences() throws InvalidConferenceFormatException {
 		try {
-			listConferencesUser = new ArrayList<>(CalendarOnline.getInstance().getOnlineConferences());
+			listConferencesUser = new ArrayList<>(new CalendarOnline(new CalDavCalendarGeneric("dav.fruux.com", "b3297431258", "jizbr5fuj9gi", "6e8c6372-eba5-43da-9eed-8e5413559c99", 443,"")).getOnlineConferences());
```
Comment (oliviercailloux) [inline src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java:126]: Concevoir un système pour ne pas inscrire le login en dur (FUTURE).

File: src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java [hunk 2]
```diff
@@ -124,7 +123,7 @@ public void display() {
 	 */
 	public void getConferences() throws InvalidConferenceFormatException {
```
Comment (oliviercailloux) [inline src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java:124]: Pourquoi cette méthode masque-t-elle le CalDAV4JException mais annonce InavlidConferenceFormatException ? (FUTURE?)

File: src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		assertEquals(conferenceVEvent.getProperty(Property.SUMMARY).getValue(), conference.getTitle());
+		assertEquals(conferenceVEvent.getProperty(Property.LOCATION).getValue(),
+				conference.getCity() + "," + conference.getCountry());
+		assertEquals(conferenceVEvent.getProperty(Property.UID).getValue(), conference.getUid());
+		assertEquals(conferenceVEvent.getProperty(Property.DESCRIPTION).getValue(),
+				"Fee:" + conference.getFeeRegistration());
+	}
+
+	@Test
+	public void testAddOnlineConference() throws MalformedURLException, URISyntaxException, ParseException,
```
Comment (oliviercailloux) [inline src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java:110]: BP Local design Exceptions

File: src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	InvalidConferenceFormatException, com.github.caldav4j.exceptions.CalDAV4JException {
+		CalendarOnline instanceCalendarOnline = new CalendarOnline(new CalDavCalendarGeneric("us.cloudamo.com", "sebastien.bourg@dauphine.eu", "600bec84476fb1", "b", 443,"/remote.php/dav"));
+		LocalDate start_ = null;
```
Comment (oliviercailloux) [inline src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java:113]: Pourquoi `start_` ? (OPT)

File: src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			start_ = LocalDate.parse("06/08/2019", formatter);
+			end_ = LocalDate.parse("08/08/2019", formatter);
+		} catch (Exception e) {
+			throw new IllegalArgumentException("Date impossible to put in the conference", e);
+		}
+		Conference conference = new Conference(uid, new URL("http://fruux.com"), "Java formation", start_, end_, 1.36,
+				"France", "Paris");
+		instanceCalendarOnline.addOnlineConference(conference);
+		Optional<Conference> confTest = instanceCalendarOnline.getConferenceFromUid(uid);
+		if(!confTest.isPresent()) {
```
Comment (oliviercailloux) [inline src/test/java/io/github/oliviercailloux/jconfs/calendar/TestCalendarOnlineNextcloud.java:127]: `assertTrue` SVP

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+	 * @param calendarID
+	 * @param port
+	 * @param postUrl
+	 */
+	public CalDavCalendarGeneric(String url, String userName, String password, String calendarID,int port,String postUrl) {
+		this.url = url;
+		this.userName = userName;
+		this.password = password;
+		this.calendarID = calendarID;
+		this.port=port;
```
[the remark this answers is not in the document]
Comment (machria) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: We think we resolve it.

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	protected String url;
+	protected String userName;
+	protected String password;
+	protected String calendarID;
+	protected CredentialsProvider credsProvider = new BasicCredentialsProvider();
+	protected CloseableHttpClient httpclient;
+	protected HttpHost hostTarget;	
+	protected CalDAVCollection collectionCalendarsOnline;
+	protected int port;
+	protected String posturl;
```
[the remark this answers is not in the document]
Comment (machria) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: done

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+ *  @author machria & sbourg
+ *  Builder which create the connection with online calendar
+ */
+public class CalDavCalendarGeneric {
+	protected String url;
+	protected String userName;
+	protected String password;
+	protected String calendarID;
```
[the remark this answers is not in the document]
Comment (machria) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: done

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	/**
+	 * Constructor for a generic calendar object 
+	 * @param url
+	 * @param userName
+	 * @param password
+	 * @param calendarID
+	 * @param port
+	 * @param postUrl
+	 */
+	public CalDavCalendarGeneric(String url, String userName, String password, String calendarID,int port,String postUrl) {
```
[the remark this answers is not in the document]
Comment (sebastienbourg) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: #7

[the remark this answers is not in the document]
Comment (sebastienbourg) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: issue create #7

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalendarOnline.java [hunk 10]
```diff
@@ -56,59 +63,40 @@
...
+		this.connector=connector;
 	}
 
+
 	/**
 	 * This method has been partially taken from :
 	 * https://github.com/dedeibel/list-events-caldav4j-example/blob/master/src/test/java/benjaminpeter/name/ListCalendarTest.java
 	 * It allows you to retrieve events from calendars hosted on a fruux account
-	 * 
+	 * Maybe add a try catch in line 97 because if an event don't have an url in vEvent it block the program.
```
[the remark this answers is not in the document]
Comment (sebastienbourg) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalendarOnline.java:82]: Issue create #9

File: src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java (hunk 1 above)
Comment (sebastienbourg) [reply src/main/java/io/github/oliviercailloux/jconfs/gui/GuiListConferences.java]: issue create #10

File: src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+		this.url = url;
+		this.userName = userName;
+		this.password = password;
+		this.calendarID = calendarID;
+		this.port=port;
+		this.posturl=postUrl;
+		this.credsProvider = new BasicCredentialsProvider();
+		credsProvider.setCredentials(new AuthScope(this.url, 443),
+				new UsernamePasswordCredentials(this.userName, this.password));
```
[the remark this answers is not in the document]
Comment (sebastienbourg) [reply src/main/java/io/github/oliviercailloux/jconfs/calendar/CalDavCalendarGeneric.java]: issue create #11

**Pull request merged**